lib/nimble_options.ex

```diff
@@ -188,9 +188,47 @@ defmodule NimbleOptions do
       ...> Exception.message(error)
       "expected :interval to be a positive integer, got: :oops! (in options [:producer, :rate_limiting])"
 
+  ## Validating Schemas
+
+  Each time `validate/2` is called, the given schema itself will be validated before validating
+  the options.
+
+  In most applications the schema will never change but validating options will be done
+  repeatedly.
+
+  To avoid the extra cost of validating the schema, it is possible to validate the schema once,
+  and then use that valid schema directly. This is done by using the `new!/1` function first, and
+  then passing the returned schema to `validate/2`.
+
+  ### Example
+
+      iex> raw_schema = [
+      ...>   hostname: [
+      ...>     required: true,
+      ...>     type: :string
+      ...>   ]
+      ...> ]
+      ...>
+      ...> schema = NimbleOptions.new!(raw_schema)
+      ...>
+      ...> config = [hostname: "elixir-lang.org"]
+      ...>
+      ...> NimbleOptions.validate(config, schema)
+      {:ok, hostname: "elixir-lang.org"}
+
+  Calling `new!/1` from a function that receives options will still validate the schema each time
+  that function is called. Declaring the schema as a module attribute is supported:
+
+      @options_schema NimbleOptions.new!([...])
+
+  This schema will be validated at compile time. Calling `docs/1` on that schema is also
+  supported.
   """
 
   alias NimbleOptions.ValidationError
+  alias __MODULE__
+
+  defstruct schema: [], schema_valid?: false
```

Comment:
We should probably remove the `schema_valid?` field. I don't think we should ever build a schema that is invalid. :)

Reply:
I pushed the change